### ISSUE TITLE
Installation Issues

### ISSUE THREAD

**Issue by thejohnha** [nextcloud/vm#1708]
Found 6 issues when installing (latest ubuntu 20.4, clean server):
1. when using an ubuntu user other than ncadmin, late in the script it states that ncadmin is currently in use, but it's not.
2. TLS installation fails because it says certbot is not found. Installing certbot before running the script with `apt install certbot` fixes this issue.
3. If installing `Fail2ban-Statuscheck` it fails, message says `Sorry but the jail 'Nextcloud' does not exist
4. early in the script, it fails to list the ubuntu archive URL. only a blank is shown.
5. If previewgenerator and logrotate are installed, I get a continuous stream of `/bin/sh: 1: cannot create /var/log/nextcloud/previewgenerator.log: Permission denied` errors. The fix for which is:
`sudo touch /var/log/nextcloud/previewgenerator.log`
`sudo chown www-data: /var/log/nextcloud/previewgenerator.log`
6. After installation is complete, if I ssh into my server and become root with `sudo -i` then it runs the installation script, which it should not do.

Thanks Daniel! @enoch85

**Comment by thejohnha** [comment]
I tested the script again today and items 1, 2, 3 and 6 appear to have been resolved. Thank you!

Regarding #2, although the script now installs certbot, it does not generate the certificate anymore. See screenshot:
![20201219_162023](https://user-images.githubusercontent.com/6325833/102700144-ea26ed00-4218-11eb-975e-d7d922778257.jpg)

Regarding #4 here is a screenshot illustrating the issue:
![20201219_160931](https://user-images.githubusercontent.com/6325833/102700154-04f96180-4219-11eb-8e68-2d0ca0c26ba9.jpg)

Regarding #5, the issue still exists.

Regarding a new matter, I see a reference to php 5.5, which is probably irrelevant, but I thought I'd mention it:
![20201219_162223](https://user-images.githubusercontent.com/6325833/102700175-235f5d00-4219-11eb-91a7-e5c7936f4aa8.jpg)

Many thanks for all your hard work, Daniel @enoch85 and @szaimen !!!

**Comment by szaimen** [comment]
Hey, great to hear, that almost all problems solved itself.

### Regarding point 2:
I have never experienced this problem. It sounds like something is wrong with your OS. Maybe an outdated kernel? I've found this: https://dausruddin.com/ubuntu-system-does-not-fully-support-snapd-cannot-mount-squashfs-image-using-squashfs/

Coud you maybe post the output of: `sudo lsb_release -a` for debugging purposes?

---
### Regarding point 4:
This is the command which works for me: https://github.com/nextcloud/vm/blob/7d35186f116d19e4fbec7a807e7e183e826f1122/lib.sh#L42
Could you maybe post the output of: `sudo cat /etc/apt/sources.list` for debugging purposes?

---
### Regarding point 5:
I've also never seen this problem. The logfile should actually get created during the previewgenerator script here:
https://github.com/nextcloud/vm/blob/7d35186f116d19e4fbec7a807e7e183e826f1122/apps/previewgenerator.sh#L230-L231
So maybe try to reinstall previewgenerator?

And when do you get this error message? During the installation of previewgenerator or at a different time?

**Comment by thejohnha** [comment]
Thanks for the followup @szaimen :-)

Regarding Point 2: It's the same machine I tested earlier when I first created this issue... I'm running the script in a container and take snapshots at the hypervisor level for each step. Although I did update the OS between when I first created the issue and yesterday when I tested it again, I think it's more likely that a commit to this repo sometime in the last two weeks caused the new issue because on the day I created this issue, the certs were created perfectly fine after I installed certbot manually... Here are the results of `sudo lsb_release -a && sudo uname -r` (thank you!):
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
5.4.78-2-pve
```

Regarding Point 4, I should mention that it correctly shows the repository URL _in the very next screen_ (sorry I should have taken a screenshot at that point too). The contents of `/etc/apt/sources.list` are:
```
deb http://archive.ubuntu.com/ubuntu focal main restricted universe multiverse
deb http://archive.ubuntu.com/ubuntu focal-updates main restricted universe multiverse
deb http://archive.ubuntu.com/ubuntu focal-security main restricted universe multiverse
```

Regarding Point 5, I get the error after everything completes. I have postfix set up to email me all errors from the box. So I get a steady stream of emails from the box every few minutes with that error message.

Thanks again @szaimen !!   :-)

**Comment by szaimen** [comment]
@thejohnha Regarding point 2: I have unfortunately no clue what could cause the snap to fail now that you've sent the output and confirmed that your OS and kernel is up-to-date. I only know that it works on any system that I've tested it on, so far. Also there was no change concerning the way how certs are generated in the last weeks, afaik.

@enoch85 We could fix Point 4 by changing the command to 
`REPO=$(grep  "^deb " /etc/apt/sources.list | grep http | awk '{print $2}' | head -1)`
What do you think?

@thejohnha Regarding point 5: As I've already shown are we creating the the previewgenerator logfile and adjusting the access rights here: https://github.com/nextcloud/vm/blob/7d35186f116d19e4fbec7a807e7e183e826f1122/apps/previewgenerator.sh#L230-L231
So I've honestly no clue what could cause this issue.

**Comment by thejohnha** [comment]
Regarding point 5, @szaimen I see those commands in the script... Does it execute the commands with elevated privileges at that point?

I will re-test the cert generation issue and report back. Thank you!

**Comment by szaimen** [comment]
> Regarding point 5, @szaimen I see those commands in the script... Does it execute the commands with elevated privileges at that point?

Yes, since you are running the complete script with elevated privileges
https://github.com/nextcloud/vm/blob/7d35186f116d19e4fbec7a807e7e183e826f1122/apps/previewgenerator.sh#L19

**Comment by thejohnha** [comment]
Hi @szaimen, regarding Point 2, I tested it again and unfortunately the cert generation failed again, but with a different error this time:

![20201223_003124](https://user-images.githubusercontent.com/6325833/102963561-63dc0680-44b7-11eb-8e7f-94396622f111.jpg)

It complains of `snapd.failure.service` and `snapd.snap-repair.service` being a disabled unit, not starting it. I did some googling and it appears that perhaps snapd was not installed correctly and re-installing is the solution, but that doesn't seem right...


Regarding Point 4, I found another place in the script where it is omitted:

![20201223_002800](https://user-images.githubusercontent.com/6325833/102964085-c1bd1e00-44b8-11eb-9ee6-1811d5ae2abb.jpg)


Regarding Point 5, unfortunately the same error regarding Preview is generated continuously...

Thanks for all your help @szaimen and @enoch85 !

**Comment by szaimen** [comment]
> regarding Point 2, I tested it again and unfortunately the cert generation failed again, but with a different error this time:
> 
> ![20201223_003124](https://user-images.githubusercontent.com/6325833/102963561-63dc0680-44b7-11eb-8e7f-94396622f111.jpg)
> 
> It complains of `snapd.failure.service` and `snapd.snap-repair.service` being a disabled unit, not starting it. I did some googling and it appears that perhaps snapd was not installed correctly and re-installing is the solution, but that doesn't seem right...

Unfortunately I don't have a different idea than what google says concerning this.

---

> Regarding Point 4, I found another place in the script where it is omitted (see end of the 3rd line):
> 
> ![20201223_002800](https://user-images.githubusercontent.com/6325833/102964085-c1bd1e00-44b8-11eb-9ee6-1811d5ae2abb.jpg)

This would be fixed with my proposal as well:

> @enoch85 We could fix Point 4 by changing the command to
> `REPO=$(grep "^deb " /etc/apt/sources.list | grep http | awk '{print $2}' | head -1)`
> What do you think?

But @enoch85 didn't answer yet.

---

> Regarding Point 5, unfortunately the same error regarding Preview is generated continuously...

As I've already said, I cannot reproduce this, unfortunately.

**Comment by szaimen** [comment]
I've created #1737 to address point 4

**Comment by enoch85** [comment]
@thejohnha Which OS is this? You haven't provided any of the info in the first post.

Thanks.

**Comment by thejohnha** [comment]
Hi @enoch85 Daniel, it's on Ubuntu 20.04. I thought I mentioned that in my first sentence. Thanks, I'll fill in the other details if you think it's necessary.

**Comment by szaimen** [comment]
Unfortunately no issue except point 4 was reproducible so I am going ahead and close this issue since that point was fixed.
@thejohnha please reopen if you can provide more info that makes those missing issues reproducible.

**Comment by thejohnha** [comment]
Thanks @szaimen I'll test it again soon, I've been seeing all your patches come across - thank you so much for all your hard work 🙏

**Comment by thejohnha** [comment]
@szaimen I'm happy to report that the certs were generated with my latest test but only after I manually installed `certbot` -- if I don't do that, the certs fail to generate.

Thank you for all your hard work!

**Comment by thejohnha** [comment]
@szaimen this does mean that the script should be updated with `apt install certbot` somewhere, right?

**Comment by szaimen** [comment]
> this does mean that the script should be updated with `apt install certbot` somewhere, right?

No, since the [snap certbot](https://github.com/nextcloud/vm/blob/6c9bb407a49ed7fa4d8fefbb7e5aeeccaafe7515/lib.sh#L721) runs fine on all installations except yours. But I am happy to hear that you've found a workaround for your installation 👍

**Comment by thejohnha** [comment]
Gotcha! thanks for pointing that line out - thank you for all your hard work. I'll investigate further and report back. 👍
